// worktop/jwt/index.mjs
// packages/worktop/src/internal/jwt.ts
import {
  base64url as J,
  decode as H
} from "worktop/base64";
import {
  asBinary as I,
  asPEM as w,
  asUTF8 as F,
  toBinary as D,
  toUTF8 as K
} from "worktop/buffer";
import {
  HMAC as j,
  sign as W,
  verify as O
} from "worktop/crypto";
var u = /* @__PURE__ */ new Error("Invalid token"), C = /* @__PURE__ */ new Error("Expired token");
function B(e) {
  return J(D(F(JSON.stringify(e))));
}
function E(e) {
  return J(D(e));
}
function R(e) {
  return JSON.parse(K(I(e)));
}
function T(e) {
  let t = e.split(".");
  if (t.length !== 3)
    throw u;
  try {
    let s = H(t[1]);
    var r = R(H(t[0])), o = R(s);
  } catch (s) {
    throw u;
  }
  return {
    header: r,
    payload: o,
    signature: t[2]
  };
}
function k(e, t, r) {
  return t.iat = t.iat || Date.now() / 1e3 | 0, t.exp == null && r != null && (t.exp = t.iat + r), e + "." + B(t);
}
function A(e, t) {
  let r = T(t), o = r.payload;
  if (!t.startsWith(e + "."))
    throw u;
  if (o.exp != null && o.exp < Date.now() / 1e3)
    throw C;
  return r;
}
function v(e, t, r) {
  let { kid: o, typ: s, header: c, ...l } = r, f = {
    ...c,
    alg: e + t,
    typ: s || "JWT"
  };
  return o != null && (f.kid = o), { header: B(f), config: l };
}
function S(e, t) {
  let r, { key: o, expires: s, ...c } = t, { header: l, config: f } = v("HS", e, c);
  return r = {
    async sign(n) {
      let i = k(l, { ...f, ...n }, s), a = await j(`SHA-${e}`, o, i);
      return i + "." + E(a);
    },
    async verify(n) {
      let i = A(l, n);
      if (await r.sign(i.payload) !== n)
        throw u;
      return i.payload;
    }
  };
}
function g(e, t) {
  let { privkey: r, pubkey: o, expires: s, ...c } = t, { header: l, config: f } = v("RS", e, c), n = "RSASSA-PKCS1-v1_5", i = {
    name: "RSASSA-PKCS1-v1_5",
    hash: `SHA-${e}`
  };
  return {
    async sign(a) {
      let y = await crypto.subtle.importKey("pkcs8", w(r), i, !1, ["sign"]), p = k(l, { ...f, ...a }, s), d = await W(n, y, p);
      return p + "." + E(d);
    },
    async verify(a) {
      let [y, p, d] = a.split("."), m = A(l, a), b = y + "." + p, x = I(H(d)), P = await crypto.subtle.importKey("spki", w(o), i, !1, ["verify"]);
      if (await O(n, P, b, x))
        return m.payload;
      throw u;
    }
  };
}
function h(e, t) {
  let { privkey: r, pubkey: o, expires: s, ...c } = t, { header: l, config: f } = v("ES", e, c), n = {
    name: "ECDSA",
    namedCurve: `P-${e}`
  }, i = {
    name: "ECDSA",
    hash: `SHA-${e}`
  };
  return {
    async sign(a) {
      let y = await crypto.subtle.importKey("pkcs8", w(r), n, !1, ["sign"]), p = k(l, { ...f, ...a }, s), d = await W(i, y, p);
      return p + "." + E(d);
    },
    async verify(a) {
      let [y, p, d] = a.split("."), m = A(l, a), b = y + "." + p, x = I(H(d)), P = await crypto.subtle.importKey("spki", w(o), n, !1, ["verify"]);
      if (await O(i, P, b, x))
        return m.payload;
      throw u;
    }
  };
}

// packages/worktop/src/jwt.ts
var V = /* @__PURE__ */ S.bind(0, "256"), X = /* @__PURE__ */ S.bind(0, "384"), _ = /* @__PURE__ */ S.bind(0, "512"), q = /* @__PURE__ */ g.bind(0, "256"), z = /* @__PURE__ */ g.bind(0, "384"), G = /* @__PURE__ */ g.bind(0, "512"), Q = /* @__PURE__ */ h.bind(0, "256"), Y = /* @__PURE__ */ h.bind(0, "384"), ee = /* @__PURE__ */ h.bind(0, "512");
export {
  Q as ES256,
  Y as ES384,
  ee as ES512,
  C as EXPIRED,
  V as HS256,
  X as HS384,
  _ as HS512,
  u as INVALID,
  q as RS256,
  z as RS384,
  G as RS512,
  T as decode
};
